File: sdk/web/component/sqljson/controller/AuthenticationController.js
var AuthenticationController = function authenticationController(props)
{
    let user = {};

    let CookieDays = 7;

    function getPrimaryWorkgroupOption(userinfo)
    {
        let optionWorkgroups = [];

        let option = userinfo.Data.Workgroups.find(x => x.Status == 1);
        optionWorkgroups.push(option);

        return optionWorkgroups;
    }

    function loginFailed()
    {
        Utils.DeleteCookie(Constant.Login);
        Utils.DeleteCookie(Constant.Language);
    }

    function prepareUser(token, onResponse)
    {
        let usr = Utils.ParseJwt(token);
        if (usr == null)
        {
            onResponse != null ? onResponse(null) : null;
            return;
        }

        let userdata = JSON.parse(usr.Data);
        usr.Data = userdata;

        usr.PrimaryWorkgroupOption = getPrimaryWorkgroupOption(usr);
        if (usr.Language != null && usr.PrimaryWorkgroupCountry != null && usr.Language != "" && usr.PrimaryWorkgroupCountry != "") 
        {
            usr.CountryLanguage = usr.Language + '-' + usr.PrimaryWorkgroupCountry;
        }
        else
        {
            usr.CountryLanguage = 'en-US';
        }

        usr.Token = "Bearer " + token;

        usr.IsInRoles = function (roles)
        {
            for (let i = 0; i < roles.length; i++)
            {
                if (typeof usr.Roles === "string")
                {
                    if (usr.Roles == usr.HostWorkgroupCode.toLowerCase().concat('.', roles[i]))
                    {
                        return true;
                    }
                }
                else
                {
                    let workgrouprole = usr.Roles.find(x => x == usr.HostWorkgroupCode.toLowerCase().concat('.', roles[i]));
                    if (workgrouprole != null)
                    {
                        return true;
                    }
                }
            }
            return false;
        };

        return usr;
    }

    user.GetUser = function ()
    {
        let encryptedsession = Utils.GetCookie(Constant.Login);
        if (encryptedsession == null || encryptedsession == "")
        {
            return null;
        }

        let compressedToken = aescrypto.DecryptFromBase64(encryptedsession, props.AesSecretKey, props.AesInitVector);
        if (compressedToken == null || compressedToken == "")
        {
            return null;
        }

        let token = GzipCompression.DecompressedText(compressedToken);
        if (token == null || token == "")
        {
            return null;
        }

        let usr = prepareUser(token);

        return usr;
    };

    user.Logout = function (onResponse)
    {
        let args = [];

        Utils.DeleteCookie(Constant.Login, props.SingleSubDomain);
        Utils.DeleteCookie(Constant.Language, props.SingleSubDomain);
        Utils.DeleteAllCookies(props.SingleSubDomain);
        localStorage.clear();
        sessionStorage.clear();
        props.Connector.Submit(args, "ardvro/component/sqljson/Authentication/Logout");

        onResponse != null ? onResponse(null) : null;
    };

    user.SaveLoginCookies = function (data)
    {
        let compressedData = GzipCompression.CompressedText(data);
        let compressedToken = aescrypto.EncryptToBase64(compressedData, props.AesSecretKey, props.AesInitVector);
        Utils.SetCookie(Constant.Login, compressedToken, CookieDays, props.SingleSubDomain);
   };

    user.LoginToken = function (onLogin, onError)
    {
        let encryptedsession = Utils.GetCookie(Constant.Login);
        if (encryptedsession == null || encryptedsession == "")
        {
            //Utils.DeleteCookie(Constant.Login);
            //Utils.DeleteCookie(Constant.Language);
            Utils.DeleteAllCookies();
            //localStorage.clear();
            if (onError != null)
            {
                onError();
            }
            return;
        }

        let compressedToken = aescrypto.DecryptFromBase64(encryptedsession, props.AesSecretKey, props.AesInitVector);
        if (compressedToken == null || compressedToken == "")
        {
            if (onError != null)
            {
                onError();
            }
            return;
        }

        let token = GzipCompression.DecompressedText(compressedToken);
        if (token == null || token == "")
        {
            if (onError != null)
            {
                onError();
            }
            return;
        }

        //MsgLoading.Show();
        let args = [token];
        props.Connector.Submit(args, "ardvro/component/sqljson/authentication/logintoken", function (box)
        {
            //MsgLoading.Close();

            if (box == null)
            {
                loginFailed();
                if (onError != null)
                {
                    onError();
                }
                document.location.href = props.WebBaseUrl;
                return;
            }

            let obj = box.Data;
            if (obj == null || obj == "")
            {
                loginFailed();
                if (onError != null)
                {
                    onError();
                }
                document.location.href = props.WebBaseUrl;
                return;
            }

            //2020-12-17 15:27 to prevent forbidden error by browser, coz browser detect that the cookies was re-written by other subdomain.
            //user.SaveLoginCookies(box.Data);

            let usr = prepareUser(box.Data, onLogin);

            if (onLogin != null)
            {
                onLogin(usr);
            }

        });
    };

    user.TwoFactorsLogin = function (username, password, code, saveLogin, onResponse)
    {
        let args = [username, password, code];
        props.Connector.Submit(args, "ardvro/component/sqljson/Authentication/TwoFactorsLogin", function (box)
        {
            if (box != null && box.Data != null && box.Data != "")
            {
                let usr = prepareUser(box.Data, onResponse);

                if (saveLogin)
                {
                    user.SaveLoginCookies(box.Data);
                    Utils.SetCookie(Constant.Language, usr.Language, CookieDays, props.SingleSubDomain);
                }

                onResponse != null ? onResponse(usr) : null;
                return;
            }

            onResponse != null ? onResponse(null) : null;

        });
    };

    user.Login = function (username, password, saveLogin, onResponse)
    {
        let args = [username, password];
        props.Connector.Submit(args, "ardvro/component/sqljson/Authentication/login", function (box)
        {
            if (box != null && box.Data != null && box.Data != "")
            {
                let usr = prepareUser(box.Data, onResponse);

                if (saveLogin)
                {
                    user.SaveLoginCookies(box.Data);
                    Utils.SetCookie(Constant.Language, usr.Language, CookieDays, props.SingleSubDomain);
                }

                onResponse != null ? onResponse(usr) : null;
                return;
            }

            onResponse != null ? onResponse(null) : null;

        });
    };

    user.SaveLoginData = function (userData)
    {
        let usr = prepareUser(userData);

        user.SaveLoginCookies(userData);
        Utils.SetCookie(Constant.Language, usr.Language, CookieDays, props.SingleSubDomain);
    }

    user.GuestLogin = function (name, email, id, onResponse)
    {
        let args = [name, email, id];
        props.Connector.Submit(args, "ardvro/component/sqljson/Authentication/GuestLogin", function (box)
        {
            if (box != null && box.Data != null && box.Data != "")
            {
                let usr = prepareUser(box.Data, onResponse);

                user.SaveLoginCookies(box.Data);
                Utils.SetCookie(Constant.Language, usr.Language, CookieDays, props.SingleSubDomain);

                onResponse != null ? onResponse(usr) : null;
                return;
            }

            onResponse != null ? onResponse(null) : null;
        });
    };

    user.ChangePassword = function (username, password, newPassword, repassword, onResponse)
    {
        let args = [username, password, newPassword, repassword];
        props.Connector.Submit(args, "ardvro/component/sqljson/Authentication/ChangePassword", function (box)
        {
            onResponse != null ? onResponse(box == null ? null : box.Data) : null;
        });
    };

    user.OtpLogin = function (code, saveLogin, onResponse)
    {
        let args = [code];
        props.Connector.Submit(args, "ardvro/component/sqljson/Authentication/OtpLogin", function (box)
        {
            if (box != null && box.Data != null && box.Data != "")
            {
                if (saveLogin)
                {
                    user.SaveLoginCookies(box.Data);
                }

                let usr = prepareUser(box.Data, onResponse);
                Utils.SetCookie(Constant.Language, usr.Language, 1, props.SingleSubDomain);

                onResponse != null ? onResponse(usr) : null;
                return;
            }

            onResponse != null ? onResponse(null) : null;
        });
    };

    user.RequestOtp = function (email, securityType, onResponse)
    {
        let args = [email, securityType];
        props.Connector.Submit(args, "ardvro/component/sqljson/Authentication/RequestOtp", function (box)
        {
            onResponse != null ? onResponse(box == null ? null : box.Data) : null;
        });
    };

    user.ResetPasswordConfirmation = function (code, password, repassword, onResponse)
    {
        let args = [code, password, repassword];
        props.Connector.Submit(args, "ardvro/component/sqljson/Authentication/ResetPasswordConfirmation", function (box)
        {
            onResponse != null ? onResponse(box == null ? null : box.Data) : null;
        });
    };

    user.ResetPassword = function (email, onResponse)
    {
        let args = [email];
        props.Connector.Submit(args, "ardvro/component/sqljson/Authentication/resetpassword", function (box)
        {
            onResponse != null ? onResponse(box == null ? null : box.Data) : null;
        });
    };

    user.ActivateUser = function (code, saveLogin, onResponse)
    {
        let args = [code];
        props.Connector.Submit(args, "ardvro/component/sqljson/Authentication/ActivateUser", function (box)
        {
            if (box != null && box.Data != null && box.Data != "")
            {
                if (saveLogin)
                {
                    user.SaveLoginCookies(box.Data);
                }

                usr = Utils.ParseJwt(box.Data);
                if (usr == null)
                {
                    onResponse != null ? onResponse(null) : null;
                    return;
                }

                usr.PrimaryWorkgroupOption = getPrimaryWorkgroupOption();
                if (usr.Language != null && usr.PrimaryWorkgroupCountry != null && usr.Language != "" && usr.PrimaryWorkgroupCountry != "") 
                {
                    usr.CountryLanguage = usr.Language + '-' + usr.PrimaryWorkgroupCountry;
                }
                else
                {
                    usr.CountryLanguage = 'en-US';
                }

                onResponse != null ? onResponse(usr) : null;
                return;
            }

            onResponse != null ? onResponse(null) : null;
        });
    };

    user.SaveUserSetting = function (name, value, onResponse)
    {
        let args = [name, value];
        props.Connector.Submit(args, "ardvro/component/sqljson/Authentication/SaveUserSetting", function (box)
        {
            if (box != null && box.Data != null)
            {
                user.LoginToken();
            }

            if (name.Name == Constant.Language && usr.Language != null)
            {
                Utils.SetCookie(Constant.Language, usr.Language, 1, props.SingleSubDomain);
            }
            
            onResponse != null ? onResponse(box == null ? null : box.Data) : null;
        });
    };

    user.GetUserSettings = function (onResponse)
    {
        let args = [];
        props.Connector.Submit(args, "ardvro/component/sqljson/Authentication/GetUserSettings", function (box)
        {
            onResponse != null ? onResponse(box == null ? null : box.Data) : null;
        });
    };

    user.Register = function (username, password, repassword, email, onResponse)
    {
        let emailobj = { Email: email };
        let args = [username, password, repassword, JSON.stringify(emailobj)];
        props.Connector.Submit(args, "ardvro/component/sqljson/Authentication/register", function (box)
        {
            onResponse != null ? onResponse(box == null ? null : box.Data) : null;
        });
    };

    user.RegisterActivateLogin = function (username, password, repassword, email, onResponse)
    {
        let emailobj = { Email: email };
        let args = [username, password, repassword, JSON.stringify(emailobj)];
        props.Connector.Submit(args, "ardvro/component/sqljson/Authentication/RegisterActivateLogin", function (box)
        {
            onResponse != null ? onResponse(box == null ? null : box.Data) : null;
        });
    };

    user.SaveUserSecurity = function (securityType, securityValue, onResponse)
    {
        let args = [securityType, securityValue];
        props.Connector.Submit(args, "ardvro/component/sqljson/Authentication/SaveUserSecurity", function (box)
        {
            if (box != null && box.Data != null)
            {
                user.LoginToken();
            }

            onResponse != null ? onResponse(box == null ? null : box.Data) : null;
        });
    };

    user.GetUserSecurities = function (onResponse)
    {
        let args = [];
        props.Connector.Submit(args, "ardvro/component/sqljson/Authentication/GetUserSecurities", function (box)
        {
            onResponse != null ? onResponse(box == null ? null : box.Data) : null;
        });
    };

    user.ActivateUserAndWorkgroup = function (activationCode,
        workgroupCode,
        workgroupName,
        countryCode,
        workgroupCity,
        workgroupZipCode,
        workgroupAddress,
        workgroupPhone,
        workgroupEmail,
        workgroupWebsite,
        workgroupIdentityNumber,
        mailServer,
        smtpPort,
        imapPort,
        mailUsername,
        mailPassword,
        workgroupNote,
        onResponse)
    {
        let args = [
            activationCode,
            workgroupCode,
            workgroupName,
            countryCode,
            workgroupCity,
            workgroupZipCode,
            workgroupAddress,
            workgroupPhone,
            workgroupEmail,
            workgroupWebsite,
            workgroupIdentityNumber,
            mailServer,
            smtpPort,
            imapPort,
            mailUsername,
            mailPassword,
            workgroupNote
        ];

        props.Connector.Submit(args, "ardvro/component/sqljson/Authentication/activateUserandworkgroup", function (box)
        {
            onResponse != null ? onResponse(box == null ? null : box.Data) : null;
        });
    };

    user.SaveWorkgroup = function (
        id,
        workgroupType,
        workgroupCode,
        workgroupName,
        countryCode,
        workgroupCity,
        workgroupZipCode,
        workgroupAddress,
        workgroupPhone,
        workgroupEmail,
        workgroupWebsite,
        workgroupIdentityNumber,
        mailServer,
        smtpPort,
        imapPort,
        mailUsername,
        mailPassword,
        workgroupNote,
        onResponse)
    {
        let args = [
            id,
            workgroupType,
            workgroupCode,
            workgroupName,
            countryCode,
            workgroupCity,
            workgroupZipCode,
            workgroupAddress,
            workgroupPhone,
            workgroupEmail,
            workgroupWebsite,
            workgroupIdentityNumber,
            mailServer,
            Number(smtpPort),
            Number(imapPort),
            mailUsername,
            mailPassword,
            workgroupNote
        ];

        props.Connector.Submit(args, "ardvro/component/sqljson/Authentication/SaveWorkgroup", function (box)
        {
            onResponse != null ? onResponse(box == null ? null : box.Data) : null;
        });
    };

    user.GetUserWorkgroups = function (onResponse)
    {
        let args = [];
        props.Connector.Submit(args, "ardvro/component/sqljson/Authentication/GetUserWorkgroups", function (box)
        {
            onResponse != null ? onResponse(box == null ? null : box.Data) : null;
        });
    }; 

    user.IsWorkgroupCodeAvailable = function (code, onResponse)
    {
        props.Connector.Submit(code, "ardvro/component/sqljson/Authentication/IsWorkgroupCodeAvailable", function (box)
        {
            onResponse != null ? onResponse(box == null ? null : box.Data) : null;
        });
    };

    user.GetUsersByWorkgroup = function (onResponse)
    {
        let args = [];
        props.Connector.Submit(args, "ardvro/component/sqljson/Authentication/GetUsersByWorkgroup", function (box)
        {
            onResponse != null ? onResponse(box == null ? null : box.Data) : null;
        });
    };

    user.GetRoles = function (onResponse)
    {
        let args = [];
        props.Connector.Submit(args, "ardvro/component/sqljson/Authentication/GetRoles", function (box)
        {
            onResponse != null ? onResponse(box == null ? null : box.Data) : null;
        });
    };

    user.GetUserRoles = function (username, workgroup, onResponse)
    {
        let args = [username, workgroup];
        props.Connector.Submit(args, "ardvro/component/sqljson/Authentication/GetUserRoles", function (box)
        {
            onResponse != null ? onResponse(box == null ? null : box.Data) : null;
        });
    };

    user.UpdateUserWorkgroupRoles = function (id, roles, onResponse)
    {
        let args = [id, roles];
        props.Connector.Submit(args, "ardvro/component/sqljson/Authentication/UpdateUserWorkgroupRoles", function (box)
        {
            onResponse != null ? onResponse(box == null ? null : box.Data) : null;
        });
    };

    user.RemoveUserWorkgroup = function (id, onResponse)
    {
        let args = [id];
        props.Connector.Submit(args, "ardvro/component/sqljson/Authentication/RemoveUserWorkgroup", function (box)
        {
            onResponse != null ? onResponse(box == null ? null : box.Data) : null;
        });
    };

    user.AddWorkgroupUsers = function (workgroupId, usernames, roles, onResponse)
    {
        let args = [workgroupId, usernames, roles];
        props.Connector.Submit(args, "ardvro/component/sqljson/Authentication/AddWorkgroupUsers", function (box)
        {
            onResponse != null ? onResponse(box == null ? null : box.Data) : null;
        });
    };

    user.SetPrimaryWorkgroup = function (workgroupName, onResponse)
    {
        let args = [workgroupName];
        props.Connector.Submit(args, "ardvro/component/sqljson/Authentication/SetPrimaryWorkgroup", function (box)
        {
            onResponse != null ? onResponse(box == null ? null : box.Data) : null;
        });
    };

    user.GetAdministrationsMenu = function (onResponse)
    {
        props.Connector.Submit([], "ardvro/component/sqljson/Authentication/GetAdministrationsMenu", function (box)
        {
            onResponse != null ? onResponse(box == null ? null : box.Data) : null;
        });
    };

    user.RegisterDbaObjectAccess = function (workgroupId, onResponse)
    {
        let args = [workgroupId];
        props.Connector.Submit(args, "ardvro/component/sqljson/Authentication/RegisterDbaObjectAccess", function (box)
        {
            onResponse != null ? onResponse(box == null ? null : box.Data) : null;
        });
    };

    user.RegisterDbaObjectAccessAllWorkgroups = function (onResponse)
    {
        let args = [];
        props.Connector.Submit(args, "ardvro/component/sqljson/Authentication/RegisterDbaObjectAccessAllWorkgroups", function (box)
        {
            onResponse != null ? onResponse(box == null ? null : box.Data) : null;
        });
    };

    user.ImportUsers = function (list, onResponse)
    {
        let args = {
            List: list
        };
        props.Connector.Submit(args, "ardvro/component/sqljson/Authentication/ImportUsers", function (box)
        {
            onResponse != null ? onResponse(box == null ? null : box.Data) : null;
        });
    };

    return user;
};

File: sdk/.net/ardvro.sdk.net/component/sqljson/controller/AuthenticationController.cs
namespace ardvro.sdk.net.component.sqljson.controller;

using System.Text.Json;
using ardvro.sdk.net.component.sqljson.db;
using ardvro.sdk.net.component.sqljson;
using ardvro.core.ext.connection;
using ardvro.core.lib.common;
using ardvro.adapter.sso.ent;
using System.IdentityModel.Tokens.Jwt;
using System.Security.Claims;
using ardvro.adapter.sso;

public class AuthenticationController
{
    private IConnection _conection;

    public AuthenticationController(IConnection connection)
    {
        _conection = connection;
    }

    public void ChangePassword(String username, String password, String newPassword, String repassword, Action<int> onResponse)
    {
        String functionName = "ardvro/component/sqljson/Authentication/ChangePassword";
        _conection.Submit(new Object[]{ username, password, newPassword, repassword }, functionName, onResponse);
    }

    public void ResetPassword(String email, Action<int> onResponse)
    {
        String functionName = "ardvro/component/sqljson/Authentication/ResetPassword";
        _conection.Submit(new Object[]{ email }, functionName, onResponse);
    }

    public void ResetPasswordConfirmation(String code, String password, String repassword, Action<int> onResponse)
    {
        String functionName = "ardvro/component/sqljson/Authentication/ResetPasswordConfirmation";
        _conection.Submit(new Object[]{ code, password, repassword }, functionName, onResponse);
    }

    public void RequestOtp(String email, String securityType, Action<String> onResponse)
    {
        String functionName = "ardvro/component/sqljson/Authentication/RequestOtp";
        _conection.Submit(new Object[]{ email, securityType }, functionName, onResponse);
    }

    public void SaveUserSetting(String name, String value, Action<int> onResponse)
    {
        String functionName = "ardvro/component/sqljson/Authentication/SaveUserSetting";
        _conection.Submit(new Object[]{ name, value }, functionName, onResponse);
    }

    public void GetUserSettings(Action<List<ssoUserSetting>> onResponse)
    {
        String functionName = "ardvro/component/sqljson/Authentication/GetUserSettings";
        _conection.Submit(new Object[]{ }, functionName, onResponse);
    }

    public void Register(String username, String password, String repassword, String email, Action<int> onResponse)
    {
        var emailObj = new Flexible();
        emailObj.Add("Email", email);
        var jsonEmail = JsonSerializer.Serialize(email);

        String functionName = "ardvro/component/sqljson/Authentication/Register";
        _conection.Submit(new Object[]{ username, password, repassword, jsonEmail }, functionName, onResponse);
    }

    public void RegisterActivateLogin(String username, String password, String repassword, String email, Action<String> onResponse)
    {
        var emailObj = new Flexible();
        emailObj.Add("Email", email);
        var jsonEmail = JsonSerializer.Serialize(email);

        String functionName = "ardvro/component/sqljson/Authentication/RegisterActivateLogin";
        _conection.Submit(new Object[]{ username, password, repassword, jsonEmail }, functionName, onResponse);
    }

    public void SaveUserSecurity(String securityType, String securityValue, Action<int> onResponse)
    {
        String functionName = "ardvro/component/sqljson/Authentication/SaveUserSecurity";
        _conection.Submit(new Object[]{ securityType, securityValue }, functionName, onResponse);
    }

    public void GetUserSecurities(Action<List<ssoUserSecurity>> onResponse)
    {
        String functionName = "ardvro/component/sqljson/Authentication/GetUserSecurities";
        _conection.Submit(new Object[]{ }, functionName, onResponse);
    }

    public void ActivateUserAndWorkgroup(
            String activationCode,
            String workgroupCode,
            String workgroupName,
            String countryCode,
            String workgroupCity,
            String workgroupZipCode,
            String workgroupAddress,
            String workgroupPhone,
            String workgroupEmail,
            String workgroupWebsite,
            String workgroupIdentityNumber,
            String mailServer,
            int smtpPort,
            int imapPort,
            String mailUsername,
            String mailPassword,
            String workgroupNote,
            Action<int> onResponse)
    {
        String functionName = "ardvro/component/sqljson/Authentication/ActivateUserAndWorkgroup";
        Object[] args = new Object[]{
                activationCode,
                workgroupCode,
                workgroupName,
                countryCode,
                workgroupCity,
                workgroupZipCode,
                workgroupAddress,
                workgroupPhone,
                workgroupEmail,
                workgroupWebsite,
                workgroupIdentityNumber,
                mailServer,
                smtpPort,
                imapPort,
                mailUsername,
                mailPassword,
                workgroupNote
        };
        _conection.Submit(args, functionName, onResponse);
    }

    public void SaveWorkgroup(
            String id,
            String workgroupType,
            String workgroupCode,
            String workgroupName,
            String countryCode,
            String workgroupCity,
            String workgroupZipCode,
            String workgroupAddress,
            String workgroupPhone,
            String workgroupEmail,
            String workgroupWebsite,
            String workgroupIdentityNumber,
            String mailServer,
            int smtpPort,
            int imapPort,
            String mailUsername,
            String mailPassword,
            String workgroupNote,
            Action<int> onResponse)
    {
        String functionName = "ardvro/component/sqljson/Authentication/SaveWorkgroup";
        Object[] args = new Object[]{
                id,
                workgroupType,
                workgroupCode,
                workgroupName,
                countryCode,
                workgroupCity,
                workgroupZipCode,
                workgroupAddress,
                workgroupPhone,
                workgroupEmail,
                workgroupWebsite,
                workgroupIdentityNumber,
                mailServer,
                smtpPort,
                imapPort,
                mailUsername,
                mailPassword,
                workgroupNote
        };
        _conection.Submit(args, functionName, onResponse);
    }

    public void IsWorkgroupCodeAvailable(String code, Action<Boolean> onResponse)
    {
        String functionName = "ardvro/component/sqljson/Authentication/IsWorkgroupCodeAvailable";
        _conection.Submit(new Object[]{ code }, functionName, onResponse);
    }

    public void GetUserWorkgroups(Action<List<ssoWorkgroup>> onResponse)
    {
        String functionName = "ardvro/component/sqljson/Authentication/GetUserWorkgroups";
        _conection.Submit(new Object[]{ }, functionName, onResponse);
    }

    public void GetUsersByWorkgroup(Action<List<ssoWorkgroup>> onResponse)
    {
        String functionName = "ardvro/component/sqljson/Authentication/GetUsersByWorkgroup";
        _conection.Submit(new Object[]{ }, functionName, onResponse);
    }

    public void GetRoles(Action<List<Flexible>> onResponse)
    {
        String functionName = "ardvro/component/sqljson/Authentication/GetRoles";
        _conection.Submit(new Object[]{ }, functionName, onResponse);
    }

    public void GetUserRoles(String username, String workgroup, Action<List<Flexible>> onResponse)
    {
        String functionName = "ardvro/component/sqljson/Authentication/GetUserRoles";
        _conection.Submit(new Object[]{ username, workgroup }, functionName, onResponse);
    }

    public void UpdateUserWorkgroupRoles(String id, String roles, Action<int> onResponse)
    {
        String functionName = "ardvro/component/sqljson/Authentication/UpdateUserWorkgroupRoles";
        _conection.Submit(new Object[]{ id, roles }, functionName, onResponse);
    }

    public void AddWorkgroupUsers(String workgroupId, String usernames, String roles, Action<String> onResponse)
    {
        String functionName = "ardvro/component/sqljson/Authentication/AddWorkgroupUsers";
        _conection.Submit(new Object[]{ workgroupId, workgroupId, usernames, roles }, functionName, onResponse);
    }

    public void SetPrimaryWorkgroup(String workgroupName, Action<int> onResponse)
    {
        String functionName = "ardvro/component/sqljson/Authentication/SetPrimaryWorkgroup";
        _conection.Submit(new Object[]{ workgroupName }, functionName, onResponse);
    }

    public void GetAdministrationsMenu(Action<List<ssoMenu>> onResponse)
    {
        String functionName = "ardvro/component/sqljson/Authentication/GetAdministrationsMenu";
        _conection.Submit(new Object[]{ }, functionName, onResponse);
    }

    public void RegisterDbaObjectAccess(int workgroupId, Action<int> onResponse)
    {
        String functionName = "ardvro/component/sqljson/Authentication/RegisterDbaObjectAccess";
        _conection.Submit(new Object[]{ workgroupId }, functionName, onResponse);
    }

    public void RegisterDbaObjectAccessAllWorkgroups(Action<int> onResponse)
    {
        String functionName = "ardvro/component/sqljson/Authentication/RegisterDbaObjectAccessAllWorkgroups";
        _conection.Submit(new Object[]{  }, functionName, onResponse);
    }

    public void ImportUsers(List<Flexible> list, Action<int> onResponse)
    {
        var args = new Flexible();
        args.Add("List", list);

        String functionName = "ardvro/component/sqljson/Authentication/ImportUsers";
        _conection.Submit(args, functionName, onResponse);
    }

    public void RemoveUserWorkgroup(String id, Action<Object> onResponse)
    {
        String functionName = "ardvro/component/sqljson/Authentication/RemoveUserWorkgroup";
        _conection.Submit(new Object[]{ id }, functionName, onResponse);
    }


    public JwtSecurityToken ReadToken(string token)
    {
        var handler = new JwtSecurityTokenHandler();
        var jsonToken = handler.ReadToken(token);
        if (jsonToken == null)
        {
            return null;
        }

        if (!(jsonToken is JwtSecurityToken tokens))
        {
            return null;
        }

        if (tokens.ValidTo < DateTime.Now)
        {
            return null;
        }

        return tokens;
    }

    public void Login(string username, string password, Action<UserData> callback)
    {
        String functionName = "ardvro/component/sqljson/Authentication/Login";

        _conection.Submit(new Object[] { username, password }, functionName, 
            (string tokenString) => 
            {
                var token = ReadToken(tokenString);
                var userDataJson = token.Claims.FirstOrDefault(x => x.Type == "Data")?.Value;
                if (!string.IsNullOrWhiteSpace(userDataJson))
                {
                    var userData = JsonSerializer.Deserialize<UserData>(userDataJson);
                    callback?.Invoke(userData);
                }
            }
         );
    }

    public void LoginToken(string token, Action<UserData> callback)
    {
        String functionName = "ardvro/component/sqljson/Authentication/LoginToken";
        _conection.Submit(new Object[] { token }, functionName, (string tokenString) =>
           {
               var token = ReadToken(tokenString);
               var userDataJson = token.Claims.FirstOrDefault(x => x.Type == "Data")?.Value;
               if (!string.IsNullOrWhiteSpace(userDataJson))
               {
                   var userData = JsonSerializer.Deserialize<UserData>(userDataJson);
                   callback?.Invoke(userData);
               }
           }
        );
    }

    public void OtpLogin(String code, bool saveLogin, Action<UserData> callback)
    {
        String functionName = "ardvro/component/sqljson/Authentication/OtpLogin";
        _conection.Submit(new Object[]{ code, saveLogin }, functionName, (string tokenString) =>
            {
                var token = ReadToken(tokenString);
                var userDataJson = token.Claims.FirstOrDefault(x => x.Type == "Data")?.Value;
                if (!string.IsNullOrWhiteSpace(userDataJson))
                {
                    var userData = JsonSerializer.Deserialize<UserData>(userDataJson);
                    callback?.Invoke(userData);
                }
            }
        );
    }

    public void GuestLogin(String name, String email, String id, Action<UserData> callback)
    {
        String functionName = "ardvro/component/sqljson/Authentication/GuestLogin";
        _conection.Submit(new Object[]{ name, email, id }, functionName, (string tokenString) =>
            {
                var token = ReadToken(tokenString);
                var userDataJson = token.Claims.FirstOrDefault(x => x.Type == "Data")?.Value;
                if (!string.IsNullOrWhiteSpace(userDataJson))
                {
                    var userData = JsonSerializer.Deserialize<UserData>(userDataJson);
                    callback?.Invoke(userData);
                }
            }
        );
    }

    public void TwoFactorsLogin(String username, String password, String code, bool saveLogin, Action<UserData> callback)
    {
        String functionName = "ardvro/component/sqljson/Authentication/TwoFactorsLogin";
        _conection.Submit(new Object[]{ username, password, code }, functionName, (string tokenString) =>
            {
                var token = ReadToken(tokenString);
                var userDataJson = token.Claims.FirstOrDefault(x => x.Type == "Data")?.Value;
                if (!string.IsNullOrWhiteSpace(userDataJson))
                {
                    var userData = JsonSerializer.Deserialize<UserData>(userDataJson);
                    callback?.Invoke(userData);
                }
            }
        );
    }

    public void ActivateUser(String code, bool saveLogin, Action<int> onResponse)
    {
        String functionName = "ardvro/component/sqljson/Authentication/ActivateUser";
        _conection.Submit(new Object[] { code, saveLogin }, functionName, onResponse);
    }

    public void Logout(Action<int> onResponse)
    {
        String functionName = "ardvro/component/sqljson/Authentication/Logout";
        _conection.Submit(new Object[]{ }, functionName, onResponse);
    }

}
